docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,299 @@
+
+Thus, the source of truth for instance scoped variables moves from the `config` package to another part of the server (likely `domain`). See [issue #30366](https://github.com/pingcap/tidb/issues/30366).
+
+Because at this stage the source of truth is now no longer the `config` package, we will also need to decide how to handle features like `INFORMATION_SCHEMA.CLUSTER_CONFIG`. If it refers to the config file, it will not necessarily reflect the current configuration of the cluster. Because every instance setting will now have a system variable name (which becomes the unified name), I recommend that we deprecate `CLUSTER_CONFIG` for TiDB. We can change `CLUSTER_CONFIG` to read from the new source of truth and maintain both for some versions to support upgrades.
+
+## Behavior Changes
+
+There are **very few** known scenarios where a system variable is *currently both* instance and global scoped, but each of these will require behavior changes. Here is a script to detect them:
+
+```
+package main
+
+import (
+	"fmt"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/pingcap/tidb/sessionctx/variable"
+)
+
+func main() {
+	for _, v := range variable.GetSysVars() {
+		if v.HasGlobalScope() && v.HasSessionScope() && (v.GetSession != nil || v.SetSession != nil) && (v.GetGlobal != nil || v.SetGlobal != nil) {
+			fmt.Printf("%s\n", v.Name)
+		}
+	}
+}
+```
+
+**Output:**
+
+```
+tidb_store_limit
+tidb_stmt_summary_internal_query
+tidb_enable_stmt_summary
+tidb_stmt_summary_max_sql_length
+tidb_stmt_summary_max_stmt_count
+tidb_capture_plan_baselines
+tidb_stmt_summary_refresh_interval
+tidb_stmt_summary_history_size
+```
+
+Changes can be grouped into the following:
+
+1. `tidb_store_limit`: This has now been converted to global-only, see [issue #30515 (merged)](https://github.com/pingcap/tidb/issues/30515).
+2. `tidb_stmt_summary_XXX`, `tidb_enable_stmt_summary` and `tidb_capture_plan_baselines` (features work together): The recommendation discussed with the feature maintainers is to convert to global only.
+
+The change to `tidb_store_limit` is unlikely to affect users, since the feature was not working correctly. However, the change to statement summary and capture plan baselines is a behavior change which might affect some users.
+
+## Documentation Changes
+
+For the purposes of user-communication we don't need to use the terminology `INSTANCE`. We will remove more of the confusion by instead referring to the difference between these as whether it "persists to [the] cluster". This also aligns closer to the syntax that users will be using. Consider the following example changes which are easier to read (the current text also doesn't actually explain that to set an `INSTANCE` variable you must use `SET SESSION`, so it actually communicates a lot more in fewer words):
```

Comment:
> For the purposes of user-communication we don't need to use the terminology `INSTANCE`.

If so, it's also kind of confusing that `[instance]` section will be introduced to the configuration file... But for now I don't have a better idea for the naming.

Reply:
We could also call it `[tidb-server]`, this would be consistent with the naming of mysql configuration files. The only part I am worried about is the dash (which MySQL treats in a special way). If it's a problem we can change it to `[server]` or something like that.